File: apps/web/src/api/__tests__/salary.test.ts
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { salaryApi, type SalaryCalculationRequest } from '../salary';

const { mockPost } = vi.hoisted(() => {
    return { mockPost: vi.fn() };
});

// Mock axios
vi.mock('axios', () => {
    return {
        default: {
            create: vi.fn(() => ({
                post: mockPost,
                get: vi.fn(),
            })),
        },
    };
});

describe('salaryApi', () => {

    beforeEach(() => {
        mockPost.mockClear();
    });

    const mockRequest: SalaryCalculationRequest = {
        rate: 50,
        rate_type: 'Hourly',
        hours_x_day: 8,
        freelance_rate: 60,
    };

    const mockResponse = {
        gross_year: '100000',
        parsed_equation: '50 * 8 * ...',
        year_tax: '20000',
        year_tax_equation: '...',
        net_year: '80000',
        net_month: '6666',
        freelance_tax: '500',
    };

    it('should calculate salary successfully', async () => {
        mockPost.mockResolvedValueOnce({ data: mockResponse });

        const result = await salaryApi.calculate(mockRequest);

        expect(mockPost).toHaveBeenCalledWith('/salary/calculate', mockRequest);
        expect(result).toEqual(mockResponse);
    });

    it('should handle errors correctly', async () => {
        const errorMessage = 'Network Error';
        mockPost.mockRejectedValueOnce(new Error(errorMessage));

        await expect(salaryApi.calculate(mockRequest)).rejects.toThrow(
            `Error calculating salary: ${errorMessage}`
        );
    });

    it('should handle non-Error objects thrown', async () => {
        mockPost.mockRejectedValueOnce('Some string error');

        await expect(salaryApi.calculate(mockRequest)).rejects.toThrow(
            'Error calculating salary: Some string error'
        );
    });
});

File: apps/backend/api/__test__/salary_test.py
from fastapi.testclient import TestClient
from main import app

client = TestClient(app)

def test_calculate_salary_api():
    response = client.post(
        "/api/salary/calculate",
        json={
            "rate": 40,
            "rate_type": "Hourly",
            "hours_x_day": 8,
            "freelance_rate": 80
        }
    )
    if response.status_code != 200:
        print(f"Error response: {response.json()}")
    assert response.status_code == 200
    data = response.json()
    assert "gross_year" in data
    assert "net_year" in data
    assert data["parsed_equation"] == "40.0 * 8.0 * 23.3 * 11"

def test_calculate_salary_api_daily():
    response = client.post(
        "/api/salary/calculate",
        json={
            "rate": 300,
            "rate_type": "Daily",
            "hours_x_day": 8,
            "freelance_rate": 80
        }
    )
    assert response.status_code == 200
    data = response.json()
    assert "gross_year" in data
